mono/metadata/superluminal.c

```diff
@@ -0,0 +1,329 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#include <config.h>
+
+#if defined (HOST_WIN32)
+#include <glib.h>
+#include <mono/metadata/assembly.h>
+#include <mono/metadata/debug-helpers.h>
+#include <mono/metadata/debug-internals.h>
+#include <mono/metadata/debug-mono-ppdb.h>
+#include <mono/metadata/mono-debug.h>
+#include <mono/metadata/profiler.h>
+#include <mono/metadata/tabledefs.h>
+#include <mono/metadata/tokentype.h>
+#include <mono/metadata/unity-utils.h>
+
+#include <string.h>
+
+#define WIN32_LEAN_AND_MEAN
+#include "Evntcons.h"
+#include "evntprov.h"
+#include "evntrace.h"
+#include "windows.h"
+
+DECLSPEC_NOINLINE __inline VOID __stdcall Private_EventControlCallback (_In_ LPCGUID SourceId, _In_ ULONG ControlCode, _In_ UCHAR Level, _In_ ULONGLONG MatchAnyKeyword, _In_ ULONGLONG MatchAllKeyword, _In_opt_ PEVENT_FILTER_DESCRIPTOR FilterData, _Inout_opt_ PVOID CallbackContext);
+#define MCGEN_PRIVATE_ENABLE_CALLBACK_V2 Private_EventControlCallback
+#include "superluminal-CLR-ETW-Generated.h"
+
+#define MAX_NUM_OFFSETS 7000
+#define ENABLE_VERBOSE_LOGGING 0
+
+static gboolean is_initialized = FALSE;
+
+#if ENABLE_VERBOSE_LOGGING
+	static void
+	sLogInternal(const char* inFormat, ...)
+	{
+		va_list args;
+		va_start (args, inFormat);
+
+		char buffer[1024] = { 0 };
+		vsprintf_s (buffer, 1024, inFormat, args);
+
+		va_end(args);
+
+		OutputDebugStringA (buffer);
+	}
+
+	#define SUPERLUMINAL_LOG_ARGS(format, ...) sLogInternal("SUPERLUMINAL: "format"\n", __VA_ARGS__)
+	#define SUPERLUMINAL_LOG(str) sLogInternal("SUPERLUMINAL: %s\n", str)
+#else
+	#define SUPERLUMINAL_LOG_ARGS(format, ...)
+	#define SUPERLUMINAL_LOG(str)
+#endif
+
+static void
+image_event (MonoImage *image, gboolean is_rundown, gboolean is_unload)
+{
+	if (!MICROSOFT_WINDOWS_DOTNETRUNTIME_PROVIDER_Context.IsEnabled && !MICROSOFT_WINDOWS_DOTNETRUNTIME_RUNDOWN_PROVIDER_Context.IsEnabled)	{
+		SUPERLUMINAL_LOG ("Providers not enabled, skipping image_event");
+		return;
+	}
+
+	const char *pdb_path = NULL;
+	guint8 pe_guid[16] = {0};
+	gint32 pe_age = 0;
+	gint32 pe_timestamp = 0;
+
+	// We only emit PDB info for loads *and* if the image is not a dynamic image (i.e. containing dynamic methods). This is becuase
+	// dynamic images do not represent an actual on-disk image and so don't have any PDB info (calling mono_ppdb_get_signature will lead to a crash)
+	if (!is_unload && !mono_image_is_dynamic (image))
+		mono_ppdb_get_signature (image, &pdb_path, pe_guid, &pe_age, &pe_timestamp);
+
+	gunichar2 *image_path_utf16 = u8to16 (mono_image_get_filename (image));
+	gunichar2 *pdb_path_utf16 = pdb_path != NULL ? u8to16 (pdb_path) : NULL;
+
+	MonoAssembly *assembly = mono_image_get_assembly (image);
+
+	if (is_rundown) {
+		EventWriteModuleDCEnd_V2 ((uint64_t)image, (uint64_t)assembly, 0, 0, image_path_utf16, L"", 0, (GUID *)pe_guid, pe_age, pdb_path_utf16 == NULL ? L"" : pdb_path_utf16, NULL, 0, L"");
```

Comment:
@joshpeterson Question: the original code passed in `&GUID_NULL` for the 11th parameter, while this code just passes in `NULL`. Was there a particular reason for that change? I'm asking because I'm not 100% sure if this is functionally identical from ETW's perspective.

(The same thing goes for the other cases in this else/if tree)

Reply:
Oh, good point, I should have pointed out the few changes. 

I made that change because `GUID_NULL` does not seem to exist in the normal Win32 API headers. It seems to be from ATL maybe? I could not find much documentation on it.

In my testing it looks like things worked - but there may be a case here that I missed.

Reply:
I just double-checked this in the generated code and I would expect this to crash if you pass in `NULL` for that argument. The argument is internally used as source ptr for what basically amounts to a `memcpy` of `sizeof(GUID)` bytes, so I'd expect that to result in an access violation if you pass null.

Reply:
Sorry, missed your previous reply. `GUID_NULL` is in <cguid.h>, but I guess there's no need to use `GUID_NULL` specifically; it's just a GUID with all members zeroed out. If you don't want to take a dependency on another header, we could just as easily add our own static 'null guid' to this file and reference that instead. I think the main thing is that it shouldn't be just a nullptr.

Reply:
Oh, thanks for checking on this. I'm fine with including cguid.h - I'll make this change then.